File: Richie/Merge K sorted list/merge-k-sorted-lists.py
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        '''
        lists = [[1,4,5],[1,3,4],[2,6]]
        
        res = None
        curr_head = None
        
        '''
        
        res = curr_head = None
        for lst in lists:
            curr_head = self.mergeTwoLists(lst, curr_head )
            
        return curr_head
        
    def mergeTwoLists(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        head = ListNode()
        res = head
        
        # 0 - 1 - 1 -2 -3 -4
       # l1_ponter = 4
      #  l2_pointer = none
        
        while l1 and l2:
            if l1.val < l2.val:
                res.next = l1
                l1 = l1.next
            else:
                res.next = l2
                l2 = l2.next
            
            res = res.next
        
        rem  =  l1 if l1 else l2
        res.next = rem
        
        
        return head.next

File: mergeKSortedLists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int n = lists.size();
	// if size of lists is zero then return zero.
    if(n == 0)
    {
        return NULL;
    }
	// initialize min heap/ min(priority queue). 
    priority_queue<int,vector<int>,greater<int>>pq;
    int i;
	
    for(i=0;i<n;i++)
    {
        ListNode* temp = lists[i];
        while(temp != NULL)
        {
            pq.push(temp->val);
            temp = temp->next;
        }
    }
  ListNode* start = new ListNode(0);
    ListNode* head = start;
    while(!pq.empty())
    {
        head->next = new ListNode(pq.top());
        pq.pop();
        head = head->next;
    }
    return start->next;
    }
};
